samples: sensor: lps22hh: replace platform filter with fixture

pabigot opened the pull request:
Use a fixture to gate device testing, rather than a platform that doesn't have the sensor unless it's externally connected.

Fixes #28059

pabigot committed:
samples: sensor: lps22hh: replace platform filter with fixture

Remove reference to a specific board, since that board doesn't have
this sensor.

Add a dependency on presence of a devicetree node that will be
required to build the driver.

Add a fixture that identifies the specific sensor required to run the
sample.  This should be added as a side effect of specifying the
shield (though this is not currently done, it must be added in the
device description).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
There are over 20 other samples that have the same issue of identifying specific boards that may have the sensor on-board, or may have a sample-specific conf/overlay that provides it even though it's not on board.

I used these rules:
* Replace `platform_allow` with a hardware fixture requirement in situations where device testing is being performed.
* Do *not* use `depends_on` which comes from the `<board>.yaml` file because that only works for boards (AFAICT shields don't have an analogous feature set)

What I'm not sure about is whether I should also:
* Use `filter` with `dt_foo()` to ensure the sensor is present.

I'm thinking probably yes to handle build testing, so this would need to be updated.

erwango reviewed (changes requested):
IIUC, for a board in a bench with x_nucleo_iks01a3 shield, in order to test this sample, board will have to declare each sensor of the shield with 'fixture_bus_sensor', is that correct ?

Doesn't it prevents the basic case of testing on boards that have a the matching sensor on board (unless they declare the fixture, but then we lose the benefit of using device tree for test selection) ?

pabigot commented:
> IIUC, for a board in a bench with x_nucleo_iks01a3 shield, in order to test this sample, board will have to declare each sensor of the shield with 'fixture_bus_sensor', is that correct ?
> 
> Doesn't it prevents the basic case of testing on boards that have a the matching sensor on board (unless they declare the fixture, but then we lose the benefit of using device tree for test selection) ?

I think the answer is "yes" and "yes" qualified by "that's how it's supposed to work", but I'll leave that to @nashif to explain if what follows isn't correct/compelling.

When built with the sample specifying a shield was present the devicetree indicates the sensor is available, but that doesn't mean the device in the test rack actually has the shield installed.  Which was the problem in #28059.  The devicetree filter keeps something from building when the driver won't compile; the fixture prevents something from running when the sensor isn't installed.

The original `sample.yaml` here was too tightly coupled to the shield, and would be wrong if a board actually did have the sensor on it.

erwango commented:
> > Doesn't it prevents the basic case of testing on boards that have a the matching sensor on board (unless they declare the fixture, but then we lose the benefit of using device tree for test selection) ?
> 
> I think the answer is "yes" and "yes" qualified by "that's how it's supposed to work", but I'll leave that to @nashif to explain if what follows isn't correct/compelling.

To me a fixture is there to describe a manual board addition (like a loopback wire or any kind of shield), not a component which is part of BSP...
Using `filter: dt_compat_...` should actually be the way to check this; either the device is there as part of default board description or added by a shield (using `extra_args`).

nashif reviewed (commented):
Well, this can be looked at in different ways and what @erwango says is correct, but this test would build on the allowed platforms however it will never be able to run because you need the shield, so the shield in this case is the "fixture". It is a hardware addition on top of the board that is not part of the BSP of the board definition but needs to be available.
You tell sanitycheck that you have this hardware addition or modification using the "fixture", otherwise the test will fail to run.

So, I would clarify this and fix the fixture name to make it clear.

This is a common theme across all sensor samples that would require an overlay, we have overlays for some boards just to get the test to build, but it will be useless for most people running tests on hardware using --device-testing because their board might not be covered with an overlay and we have no way of adding overlays for each board we support, so the shield, whether one that is actually exists or virtual shields that take care of the DTS part without having to add an overlay for each board in each sample is the fixture solution.

pabigot commented:
OK, I've put the shield back, and updated the description in the commit message to explain the what and why based on my understanding of what the sample policy is.

> Remove reference to a specific board, since that board doesn't have this sensor.
> 
> For the purposes of the automated test retain the specification of a shield that supplies this sensor.
> 
> Add a dependency on presence of a devicetree node that will be required to build the driver.  This is added as a side effect of specifying the shield.
> 
> Add a fixture that identifies the specific sensor required to run the sample.  This should be added as a side effect of specifying the shield (though this is not currently done, it must be added in the device description).

pabigot commented:
OK, so this is going to take more work.  Keeping the shield without a platform filter breaks everything, because most platforms don't have the node label required to build the shield devicetree.

Also, testing this with the actual hardware the test is filtered as not runnable on device.

pabigot commented:
OK, so here's where we stand:

* Took out the `platform_allow` because I think we all agree it's bad.

* Added the filter `dt_compat_enabled("st,lps22hh")` so this builds only if the board or test configuration provides the sensor that will be tested.  This resolves #28059, which is the whole reason we started down this goat trail.

* Took out the extra_args `SHIELD=x_nucleo_iks01a3` because without some sort of filter the sample gets built on all boards, and most of them won't work because the shield requires an `arduino_i2c` node label.  We don't have infrastructure to filter on presence of a node label.

* Added the fixture `fixture_i2c_lps22hh` since it specifies the sensor that's expected by the test, providing a way for the test to be run on hardware someday.

The test will be built by CI because `sensortile_box` does have this sensor on it, so we lucked out there.  No solution for sensors that aren't on any board.

Any other requirements?

erwango commented:
> OK, so here's where we stand:
> 
> * Took out the `platform_allow` because I think we all agree it's bad.
> * Added the filter `dt_compat_enabled("st,lps22hh")` so this builds only if the board or test configuration provides the sensor that will be tested.  This resolves #28059, which is the whole reason we started down this goat trail.
> * Took out the extra_args `SHIELD=x_nucleo_iks01a3` because without some sort of filter the sample gets built on all boards, and most of them won't work because the shield requires an `arduino_i2c` node label.  We don't have infrastructure to filter on presence of a node label.

In yaml files, `arduino_i2c` exists as a value of `supported:` category. It could be used as value for `depends on:`.

> * Added the fixture `fixture_i2c_lps22hh` since it specifies the sensor that's expected by the test, providing a way for the test to be run on hardware someday.
> 

What about having two tests configs:
- A BSP config: using `filter: dt_compat_enabled("st,lps22hh")` that would allow driver to be built in CI and test on board with matching BSP.
- A shield config: using `depends on: arduino_i2c` for filtering, adding ` extra_args: SHIELD=x_nucleo_iks01a3` and `fixture: fixture_shield_x_nucleo_iks01a3` to allow test on bench.

pabigot commented:
How about we get a patch (this one) in that fixes #28059 so Nordic's hardware testing doesn't fail because a board doesn't have the hardware, then in 2.5 start figuring out how to resolve the conflicting requirements of build without real hardware but run only with real hardware, and develop the tooling necessary to support the new solution?

Because [as I noted]( https://github.com/zephyrproject-rtos/zephyr/pull/28070#issuecomment-687221056) a similar problem exists multiple places, and that isn't going to get fixed for 2.4.

At least not by me.  I don't have time to keep trying out new ideas and discovering that they don't work for various reasons.

erwango commented:
> How about we get a patch (this one) in that fixes #28059 so Nordic's hardware testing doesn't fail because a board doesn't have the hardware, then in 2.5 start figuring out how to resolve the conflicting requirements of build without real hardware but run only with real hardware, and develop the tooling necessary to support the new solution?
> 
> Because [as I noted](https://github.com/zephyrproject-rtos/zephyr/pull/28070#issuecomment-687221056) a similar problem exists multiple places, and that isn't going to get fixed for 2.4.
> 
> At least not by me. I don't have time to keep trying out new ideas and discovering that they don't work for various reasons.

I'm fine with this temp fix, as long as we make sure this is identified as such and we're sure it will be fixed in V2.5.0 (and not propagated)

pabigot commented:
I don't really think this is so much a temporary fix as it is aligning this test with the way it's done more commonly already.

I could open an issue to clean up the tests for 2.5, but what will be its completion criteria and who's going to work it?  I'd planned to do more cleanup, but at this point I don't see agreement on a technical approach that handles all the use cases, especially since @nashif has plans for removing things like `depends_on` that you were proposing to use instead.

erwango commented:
> tests for 2.5, but what will be its completion criteria and who's going to work it? I'd planned to do more cleanup, but at this point I don't see agreement on a technical approach that handles all the use cases, especially since @nashif has plans for removing things like `depends_on` that you were proposing to use instead.

Ok, I was not aware of the `depends on:` removal plan.
Though, I don't think this should be the way forward as it limits testing possibilities.

PerMac commented:
When checked out on this PR the sample is filtered out so it solves the issue from #28059. The solution looks good to me, but I will refrain from the review as I don't have enough experience in this area.

carlescufi commented:
@nashif and @erwango please take another look

erwango commented:
> I could open an issue to clean up the tests for 2.5, but what will be its completion criteria and who's going to work it? I'd planned to do more cleanup, but at this point I don't see agreement on a technical approach that handles all the use cases, especially since @nashif has plans for removing things like depends_on that you were proposing to use instead.

I can't find the related issue/PR, but I read a dt related discussion agreeing on the concept on providing tools to filter on alternate node label (such as `arduino_i2c`). This would allow a to solve the issue I see in this fix. Based on this promise, I'm accepting the change.
